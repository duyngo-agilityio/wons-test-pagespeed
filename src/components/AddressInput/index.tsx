'use client';

import { ChangeEvent, useEffect, useState } from 'react';
import { InputProps } from '@nextui-org/react';
import { useDebounce } from 'use-debounce';

// Icons
import { HiLocationMarker } from 'react-icons/hi';

// Components
import { Input, Text } from '@/components';
import { getLocationSuggestion } from '@/services';

interface AddressInputProps extends InputProps {}

interface LocationItem {
  properties: {
    formatted: string;
    name: string;
  };
}

const AddressInput = ({ placeholder = ' ', ...rest }: AddressInputProps) => {
  const [searchAddress, setSearchAddress] = useState('');
  const [locationsSuggestion, setLocationsSuggestion] = useState<
    LocationItem[]
  >([]);

  // Debounce searchAddress with a delay of 500ms
  const [debouncedSearchAddress] = useDebounce(searchAddress, 500);

  const handleOnChangeSearch = (event: ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setSearchAddress(value);
  };

  useEffect(() => {
    const handleFetchData = async (keyword: string) => {
      const locationResponse = await getLocationSuggestion(keyword);

      setLocationsSuggestion(
        locationResponse?.features ? locationResponse.features : [],
      );
    };

    if (debouncedSearchAddress) {
      handleFetchData(debouncedSearchAddress);
    } else {
      setLocationsSuggestion([]);
    }
  }, [debouncedSearchAddress]);

  const handleClickSuggestion = (value: string) => {
    setSearchAddress(value);
    setLocationsSuggestion([]);
  };

  return (
    <div className="relative flex-1">
      <Input
        placeholder={placeholder}
        value={searchAddress}
        onChange={handleOnChangeSearch}
        endContent={<HiLocationMarker />}
        {...rest}
      />
      {locationsSuggestion.length !== 0 && (
        <div className="shadow-md p-3 absolute z-50 bg-white dark:bg-gray-400 w-full">
          {locationsSuggestion.map(({ properties }: LocationItem) => {
            const { formatted, name } = properties;

            return (
              <button
                key={name}
                className="flex w-full text-left py-[10px] px-[4px] gap-4 items-center relative cursor-pointer dark:text-white/80"
                onClick={() => handleClickSuggestion(formatted)}
              >
                <div className="w-[16px] h-[16px]">
                  <HiLocationMarker />
                </div>

                <Text text={formatted} className="dark:text-white/80" />
              </button>
            );
          })}
        </div>
      )}
    </div>
  );
};

export default AddressInput;
