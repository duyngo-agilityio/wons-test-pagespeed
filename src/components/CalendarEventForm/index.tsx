'use client';

import { memo, useState, useCallback, useEffect } from 'react';
import { Controller, useForm } from 'react-hook-form';
import isEqual from 'react-fast-compare';
import dayjs from 'dayjs';

import { Select, SelectItem } from '@nextui-org/react';

// Constants
import { MESSAGES } from '@/constants';

// Types
import { TEventResponse } from '@/types';

// Utils
import {
  capitalizeFirstLetter,
  clearErrorOnChange,
  formatDateString,
  formatDateToISO,
  formatEventDate,
  formatToCalendarDate,
  formatToStandardDate,
  getUserIds,
  parseStringToNumberArray,
} from '@/utils';

// Api
import { getUsers } from '@/api'; // Import API to fetch users

// Models
import { IEvent, TUser } from '@/models';

// Components
import {
  ClockIcon,
  CalendarIcon,
  LocationIcon,
  PeopleIcon,
  Input,
  Button,
  Text,
  DateTimePickerModal,
  AddressInput,
} from '@/components';

interface TimeRangeProps {
  start: string;
  end: string;
}

interface EventForm {
  title: string;
  location?: string;
  people?: string | number[];
}

interface CalendarEventFormProps {
  title: string;
  previewData?: EventForm | null;
  timeRange: TimeRangeProps;
  eventTitle: string;
  date: Date;
  user: TUser;
  repeatSetting?: string;
  onSubmit: (data: Partial<IEvent>) => void;
  onClose: () => void;
}

const CalendarEventForm = ({
  previewData,
  title = '',
  timeRange,
  eventTitle,
  date,
  user,
  repeatSetting = 'does not repeat',
  onClose,
  onSubmit,
}: CalendarEventFormProps) => {
  const [isUserListOpen, setIsUserListOpen] = useState(false);
  const [users, setUsers] = useState<TUser[]>([]); // State to store user list
  const [isDateTimePickerOpen, setIsDateTimePickerOpen] = useState(false);
  const [calendarDate, setCalendarDate] = useState(formatToCalendarDate(date));
  const [startTime, setStartTime] = useState(timeRange.start);
  const [endTime, setEndTime] = useState(timeRange.end);

  const {
    handleSubmit,
    control,
    clearErrors,
    reset,
    formState: { errors, isValid },
  } = useForm<EventForm>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    defaultValues: previewData
      ? {
          title: previewData?.title || '',
          location: previewData?.location,
          people: getUserIds(previewData as TEventResponse) || '',
        }
      : {
          title: eventTitle,
          location: '',
          people: '',
        },
  });

  const standardDate = formatToStandardDate(calendarDate);
  const formattedDate = formatEventDate(standardDate);

  const toggleDateTimePicker = useCallback(() => {
    setIsDateTimePickerOpen((prev) => !prev);
  }, [setIsDateTimePickerOpen]);

  const handleFormSubmit = handleSubmit((data) => {
    const people = parseStringToNumberArray(data.people as string);

    const formattedStart = formatDateToISO(
      new Date(formatDateString(calendarDate)),
      dayjs(startTime, 'hh:mma').utc(true).format('HH:mm'),
    );
    const formattedEnd = formatDateToISO(
      new Date(formatDateString(calendarDate)),
      dayjs(endTime, 'hh:mma').utc(true).format('HH:mm'),
    );

    onSubmit({
      ...data,
      users_permissions_users: [Number(user.id), ...people],
      startTime: formattedStart,
      endTime: formattedEnd,
      date: new Date(formatDateString(calendarDate)),
    });
  });

  const handleModalClose = () => {
    reset({ title });
    onClose();
  };

  const toggleUserList = useCallback(async () => {
    const response = await getUsers(); // Call API to fetch users

    if (response) {
      setUsers(response);
    } else {
      console.error('No data found in the response');
    }
  }, []);

  useEffect(() => {
    toggleUserList(); // Automatically fetch users when component mounts
  }, [toggleUserList]);

  const [isOpenLocation, setIsOpenLocation] = useState(false);

  const usersOptions = users
    .filter((u) => u.id.toString() !== user.id.toString())
    .map((user) => ({
      label: user.username,
      key: user.id,
    }));

  return (
    <>
      <form onSubmit={handleFormSubmit}>
        <Controller
          name="title"
          control={control}
          rules={{
            required: MESSAGES.ERROR.FIELD_REQUIRED,
          }}
          render={({
            field: { name, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Input
              classNames={{
                inputWrapper: 'w-full h-[42px] rounded-[10px] bg-gray-200/30',
                input:
                  'placeholder:text-blue-800 placeholder:opacity-20 placeholder:font-medium placeholder:text-[16px]',
              }}
              type="text"
              isInvalid={!!error}
              placeholder="Add title"
              errorMessage={error?.message}
              onChange={(e) => {
                onChange(e.target.value);
                clearErrorOnChange(name, errors, clearErrors);
              }}
              {...rest}
            />
          )}
        />

        <div
          className="flex mt-[25px] gap-[0_15px]"
          onClick={toggleDateTimePicker}
        >
          <Button
            className="!bg-pink-50 dark:!bg-pink-600 text-pink-500 dark:text-pink-500 border-none rounded-full w-10 h-10 flex justify-center items-center cursor-pointer !px-0"
            data-testid="time-button"
            onClick={toggleDateTimePicker}
          >
            <ClockIcon />
          </Button>

          <div>
            <div className="flex gap-1">
              <Text
                className="text-blue-800 text-[12px] font-normal leading-normal col-span-1"
                text={`${formattedDate}`} // Display date
              />
              <Text as="span" text="-" />
              <Text
                className="text-blue-800 text-[12px] font-normal leading-normal col-span-1 uppercase"
                text={`${startTime} - ${endTime}`} // Display time
              />
            </div>

            <Text
              className="text-[rgba(1, 13, 28, 0.50)] text-opacity-50 text-[12px] font-normal leading-normal col-span-2"
              text={`Time zone - ${capitalizeFirstLetter(repeatSetting)}`}
            />
          </div>
        </div>

        {(isUserListOpen || previewData) && (
          <Controller
            name="people"
            control={control}
            render={({ field: { onChange, value, name, ...rest } }) => (
              <Select
                selectionMode="multiple"
                label="Add People"
                defaultSelectedKeys={value}
                placeholder=" "
                labelPlacement="outside"
                variant="flat"
                classNames={{
                  trigger:
                    'w-full bg-gray-50 dark:bg-gray-600 hover:!bg-gray-200/50 dark:hover:!bg-gray-900 focus:bg-gray-50 dark:focus:bg-gray-600 py-[26px] mt-5',
                  label: 'text-xl font-medium pb-1',
                }}
                onChange={(e) => {
                  onChange(e.target.value);
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              >
                {usersOptions.map((option) => (
                  <SelectItem key={option.key}>{option.label}</SelectItem>
                ))}
              </Select>
            )}
          />
        )}

        {(isOpenLocation || previewData?.location) && (
          <Controller
            name="location"
            control={control}
            render={({
              field: { name, onChange, ...rest },
              fieldState: { error },
            }) => (
              <AddressInput
                isInvalid={!!error}
                errorMessage={error?.message}
                label="Location"
                classNames={{ mainWrapper: 'mt-5' }}
                onChange={(value) => {
                  onChange(value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
            )}
          />
        )}

        <div className="flex gap-[0_25px] m-[30px_0]">
          {!isUserListOpen && !previewData && (
            <Button
              color="primary"
              startContent={<PeopleIcon />}
              className="text-[15px] font-medium w-auto  py-[10px] px-[25px] mt-10 mt-0"
              onClick={() => setIsUserListOpen((prev) => !prev)}
            >
              Add People
            </Button>
          )}

          {!isOpenLocation && !previewData?.location && (
            <Button
              startContent={<LocationIcon />}
              className="!bg-white font-medium dark:!bg-white text-center !text-blue-500 dark:text-white/70 border border-[1px] border-[rgba(58, 54, 219, 0.1)] py-[10px] px-[25px] !rounded-[10px] font-DM-Sans text-[14.22px] font-normal leading-normal"
              onClick={() => setIsOpenLocation(true)}
            >
              Add Location
            </Button>
          )}
        </div>

        <div className="m-[0_0_40px] flex gap-[0_15px]">
          <Button
            className="!bg-pink-50 dark:!bg-pink-600 text-pink-500 dark:text-pink-500 border-none rounded-full w-10 h-10 flex justify-center items-center cursor-pointer !px-0"
            data-testid="calendar-button"
          >
            <CalendarIcon width={15} height={15} color="pink-500" />
          </Button>

          <div className="flex flex-col gap-[0_30px]">
            <Text
              className="text-blue-800 text-[12px] font-normal leading-normal col-span-1"
              text={user.fullName} // Display name
            />
            <Text
              className="text-[rgba(1, 13, 28, 0.50)] text-opacity-50 text-[12px] font-normal leading-normal"
              text="Busy - Default visibility - notify 30 minutes before" // Display status
            />
          </div>
        </div>

        <div className="flex flex-row-reverse gap-[0_20px]">
          <Button
            className="min-w-[93px] text-[15px] font-normal w-auto py-[10px] mt-10 mt-0"
            type="submit"
            color="primary"
            isDisabled={!isValid}
          >
            Save
          </Button>

          <Button
            className="min-w-[93px] !bg-white font-normal dark:!bg-white text-center !text-blue-500 dark:text-white/70 border border-[1px] border-[rgba(58, 54, 219, 0.1)] py-[10px] !rounded-[10px] font-DM-Sans text-[15px] font-normal leading-normal"
            onClick={handleModalClose}
          >
            Close
          </Button>
        </div>
      </form>

      {/* DateTimePickerModal */}
      {isDateTimePickerOpen && (
        <DateTimePickerModal
          isOpen={isDateTimePickerOpen}
          onClose={toggleDateTimePicker}
          selectedDate={formatDateString(calendarDate)}
          selectedStartTime={startTime}
          selectedEndTime={endTime}
          onDateChange={(dateString) =>
            setCalendarDate(formatToCalendarDate(new Date(dateString)))
          }
          onStartTimeChange={setStartTime}
          onEndTimeChange={setEndTime}
        />
      )}
    </>
  );
};

export default memo(CalendarEventForm, isEqual);
