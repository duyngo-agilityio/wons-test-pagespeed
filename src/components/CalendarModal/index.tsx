'use client';

import { memo, useState } from 'react';
import { CalendarDate } from '@internationalized/date';
import { IoClose } from 'react-icons/io5';
import isEqual from 'react-fast-compare';

import { Modal as NextModal, ModalContent, Calendar } from '@nextui-org/react';

// Utils
import { formatToCalendarDate } from '@/utils';

// Models
import { IEvent, TUser } from '@/models';

// Components
import {
  Heading,
  Button,
  Tabs,
  CalendarEventForm,
  CalendarTaskForm,
} from '@/components';

interface TimeRangeProps {
  start: string;
  end: string;
}

interface EventForm {
  title: string;
  location?: string;
  people?: string | number[];
}

interface TaskForm {
  title: string;
  description: string;
}

interface EventFormModalProps {
  title: string;
  eventTitle: string;
  date: Date;
  timeRange?: TimeRangeProps;
  time?: string;
  isOpen: boolean;
  repeatSetting?: string;
  guests?: string[];
  location?: string;
  user?: TUser;
  previewData?: EventForm | TaskForm | null;
  isEdit?: boolean;
  isTask?: boolean;
  setIsTask?: (value: boolean) => void;
  onSubmit: (data: Partial<IEvent>) => void;
  onClose: () => void;
}

const CalendarModal = ({
  title = '',
  eventTitle = '',
  date,
  timeRange,
  isEdit = false,
  isTask = false,
  time,
  isOpen,
  user,
  previewData = null,
  setIsTask,
  onClose,
  onSubmit,
}: EventFormModalProps): JSX.Element => {
  const [isOpenCalendar, setIsOpenCalendar] = useState(false);
  const [calendarDate, setCalendarDate] = useState(formatToCalendarDate(date));

  const handleDateChange = (newDate: CalendarDate) => {
    setIsOpenCalendar(false);
    setCalendarDate(newDate);
  };

  const handleModalClose = () => {
    onClose();
  };

  const handleSelectionChange = (key: React.Key) => {
    if (key === 'task') {
      setIsTask && setIsTask(true);
    }
  };

  const calendarTabs = [
    {
      key: 'event',
      label: 'Event',
      content: (
        <CalendarEventForm
          previewData={previewData}
          timeRange={timeRange}
          eventTitle={eventTitle}
          date={date}
          user={user}
          onSubmit={onSubmit}
          onClose={onClose}
        />
      ),
      isDisable: isEdit && isTask,
    },
    {
      key: 'task',
      label: 'Task',
      content: (
        <CalendarTaskForm
          previewData={previewData}
          date={date}
          onClose={onClose}
          time={time ?? ''}
          onSubmit={onSubmit}
        />
      ),
      isDisable: isEdit && !isTask,
    },
    {
      key: 'reminder',
      label: 'Reminder',
      content: 'This is the reminder content',
      isDisable: true,
    },
  ];

  return (
    <NextModal
      className="max-w-[467px]"
      isOpen={isOpen}
      size="xl"
      hideCloseButton={true}
      placement="center"
    >
      <ModalContent className="relative top-0 left-0 p-[30px_30px_40px] bg-white dark:bg-gray-800 rounded-[10px] shadow-[-14px_30px_20px_0px_rgba(0,0,0,0.05)] w-[467px]">
        {isOpenCalendar && (
          <Calendar
            className="absolute top-[240px] left-[80px] z-10"
            aria-label="Date"
            value={calendarDate}
            onChange={handleDateChange}
          />
        )}

        <Heading className="mt-[12px]" size="md" title={title} />

        <Button
          className="absolute top-[30px] right-[30px] bg-pink-50 dark:bg-pink-600 text-pink-500 dark:text-pink-500 border-none rounded-full w-10 h-10 flex justify-center items-center cursor-pointer px-0"
          data-testid="close-button"
          onClick={handleModalClose}
        >
          <IoClose size={20} />
        </Button>

        <Tabs
          style={{ padding: 0, margin: '40px 0 18px' }}
          tabs={calendarTabs}
          defaultSelectedKey={isEdit && isTask ? 'task' : 'event'}
          onSelectionChange={handleSelectionChange}
        />
      </ModalContent>
    </NextModal>
  );
};

export default memo(CalendarModal, isEqual);
