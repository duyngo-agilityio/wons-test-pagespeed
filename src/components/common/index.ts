// Custom-icons
export * from './icons';

// Table component
export * from './Table';

// React-icons
export { CiSearch } from 'react-icons/ci';
export { FaCalendarAlt } from 'react-icons/fa';
export { FaChevronLeft } from 'react-icons/fa6';
export { FcGoogle } from 'react-icons/fc';
export { FaFacebookF } from 'react-icons/fa';
export { IoCheckmarkOutline } from 'react-icons/io5';
export { IoLogOut } from 'react-icons/io5';
export { TbDots } from 'react-icons/tb';
export { FiSun } from 'react-icons/fi';
export { IoMoonOutline } from 'react-icons/io5';
export { IoIosClose } from 'react-icons/io';

export { default as Button } from './Button';
export { default as Card } from './Card';
export { default as Checkbox } from './Checkbox';
export { default as DatePicker } from './DatePicker';
export { default as DateRangePicker } from './DateRangePicker';
export { default as DropdownActions } from '../DropdownActions';
export { default as Heading } from './Heading';
export { default as Image } from './ImageFallback';
export { default as ImageFallback } from './ImageFallback';
export { default as Input } from './Input';
export { default as LoadingIndicator } from './LoadingIndicator';
export { default as NoProductAvailable } from './Message';
export { default as SearchInput } from './SearchInput';
export { default as Spinner } from './Spinner';
export { default as Text } from './Text';
export { default as Toast } from './Toast';
export * from '../Skeleton';
export { default as Autocomplete } from './Autocomplete';
