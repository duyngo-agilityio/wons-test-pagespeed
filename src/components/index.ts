export * from './common';

export { default as ErrorBoundary } from './ErrorBoundary';
export { default as RecentServicesTable } from './RecentServicesTable';
export { default as SignInForm } from './SignInForm';
export { default as SignUpForm } from './SignUpForm';
export { default as StatisticCard } from './StatisticCard';
export { default as ProductCard } from './ProductCard';
export { default as Logo } from './Logo';
export { default as ThemeSwitcher } from './ThemeSwitcher';
export { default as ToastList } from './ToastList';
export { default as AddressInput } from './AddressInput';
