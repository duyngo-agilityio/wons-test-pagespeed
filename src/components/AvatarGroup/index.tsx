import { memo } from 'react';
import isEqual from 'react-fast-compare';
import clsx from 'clsx';

// components
import { Button, ImageFallback } from '@/components';

// models
import { TUser } from '@/models';

// icons
import { FiPlus } from 'react-icons/fi';

// Types
import { StrapiModel } from '@/types';

// Utils
import { getSubarray } from '@/utils';

type AvatarGroupProps = {
  users: StrapiModel<Omit<TUser, 'id'>>[];
};

const AvatarGroup = ({ users }: AvatarGroupProps) => (
  <div className="flex items-center">
    {getSubarray<StrapiModel<Omit<TUser, 'id'>>>(users, 0, 4).map(
      ({ id, attributes }, index) => {
        const { avatar = '', username = '' } = attributes ?? {};

        return (
          <div key={id} className={clsx('w-7 h-7', index !== 0 && '-ml-3')}>
            <ImageFallback
              src={avatar}
              alt={username}
              placeholder={null}
              width={28}
              height={28}
              className="h-full rounded-full object-cover"
            />
          </div>
        );
      },
    )}

    <Button
      isIconOnly
      color="success"
      aria-label="Plus Button"
      className="w-7 h-7 p-0 -ml-3 rounded-full border-0"
    >
      <FiPlus />
    </Button>
  </div>
);

export default memo(AvatarGroup, isEqual);
