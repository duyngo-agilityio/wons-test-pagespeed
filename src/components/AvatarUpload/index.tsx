'use client';

import { ChangeEvent, useCallback, useState } from 'react';

// icons
import { IoCamera } from 'react-icons/io5';

import { ImageFallback, Input } from '@/components';

// constants
import { ERROR_MESSAGES, MAX_SIZE, REGEX } from '@/constants';

export type TUpdateProfileProps = {
  onFileChange: (file: File) => void;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
  value: string;
  error?: string;
};

const AvatarUpload = ({
  value,
  error = '',
  onChange,
  onFileChange,
}: TUpdateProfileProps) => {
  const [previewURL, setPreviewURL] = useState<string>('');
  const [errorMessage, setErrorMessage] = useState<string>(error);

  const handleChangeFile = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      const file = (e.target.files && e.target.files[0]) as File;

      if (!file) {
        return;
      }

      if (!REGEX.IMG.test(file.name)) {
        const error = ERROR_MESSAGES.UPLOAD_IMAGE;
        setErrorMessage(error);
        return;
      }

      if (file.size > MAX_SIZE) {
        const error = ERROR_MESSAGES.UPLOAD_IMAGE_SIZE;
        setErrorMessage(error);
        return;
      }

      setErrorMessage('');

      if (previewURL) {
        URL.revokeObjectURL(previewURL);
        setPreviewURL('');
      }

      const previewImage = URL.createObjectURL(file);
      setPreviewURL(previewImage);
      onFileChange(file);
    },
    [onFileChange, previewURL],
  );

  const handleOnchange = (e: ChangeEvent<HTMLInputElement>) => {
    handleChangeFile(e);
    onChange(e);
  };

  return (
    <div className="!m-0 flex flex-col justify-center items-center my-20">
      <label
        htmlFor="file"
        className="cursor-pointer hover:scale-110 transition-transform"
      >
        <div className="rounded-full w-32 h-32 bg-gray-50 dark:bg-gray-600 flex justify-center items-center">
          {previewURL || value ? (
            <ImageFallback
              src={previewURL || value}
              alt="Avatar"
              width={128}
              height={128}
              className="rounded-full object-cover w-full h-full"
            />
          ) : (
            <IoCamera
              size={32}
              className="text-blue-800/70 dark:text-white/70"
            />
          )}
        </div>
      </label>

      <Input
        aria-label="Upload Avatar"
        type="file"
        id="file"
        className="hidden"
        accept="image/*"
        onChange={handleOnchange}
        isInvalid={!errorMessage}
        errorMessage={errorMessage}
      />

      {(errorMessage || error) && (
        <p className="text-red-500 text-md mt-2 z-100">
          {errorMessage || error}
        </p>
      )}
    </div>
  );
};

export default AvatarUpload;
