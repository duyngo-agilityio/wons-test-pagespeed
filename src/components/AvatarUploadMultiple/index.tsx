'use client';

import { ChangeEvent, useCallback, useState } from 'react';
import { IoCamera, IoClose } from 'react-icons/io5';
import { Input, Button, Text, ImageFallback } from '@/components';

// constants
import { ERROR_MESSAGES, MAX_SIZE, REGEX } from '@/constants';

export type TAvatarUploadMultipleProps = {
  onFileChange: (files: File[]) => void;
  value?: string[];
  error?: string;
};

const AvatarUploadMultiple = ({
  value = [],
  error = '',
  onFileChange,
}: TAvatarUploadMultipleProps) => {
  const [previewURLs, setPreviewURLs] = useState<string[]>(value);
  const [errorMessage, setErrorMessage] = useState<string>(error);
  const [_, setFilesList] = useState<File[]>([]);

  const handleChangeFiles = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      const files = e.target.files;
      if (!files) return;

      // Reset error message
      setErrorMessage('');

      if (files.length > 2) {
        setErrorMessage(ERROR_MESSAGES.MAX_IMAGE);
        return;
      }

      // Array to hold valid files and preview URLs
      const validFiles: File[] = [];
      const newPreviewURLs: string[] = [];

      // Validate files
      for (let i = 0; i < files.length; i++) {
        const file = files[i];

        // Validate file name and size
        if (!REGEX.IMG.test(file.name)) {
          setErrorMessage(ERROR_MESSAGES.UPLOAD_IMAGE);
          return;
        }
        if (file.size > MAX_SIZE) {
          setErrorMessage(ERROR_MESSAGES.UPLOAD_IMAGE_SIZE);
          return;
        }

        // Add valid file
        validFiles.push(file);
        const previewImage = URL.createObjectURL(file);
        newPreviewURLs.push(previewImage);
      }

      // Use the callback version of setState to ensure we have the latest value
      setFilesList((prevFiles) => {
        const updatedFiles = [...prevFiles, ...validFiles];

        // Pass the updated files list to the parent component
        onFileChange(updatedFiles);

        return updatedFiles; // Update the filesList state
      });

      // Update preview URLs
      setPreviewURLs((prevURLs) => [...prevURLs, ...newPreviewURLs]);
    },
    [onFileChange], // Remove filesList from dependency array
  );

  const handleRemoveImage = (index: number) => {
    // Remove the image from previewURLs
    const updatedPreviews = previewURLs.filter((_, i) => i !== index);
    setPreviewURLs(updatedPreviews);

    // Also remove the corresponding file from filesList
    setFilesList((prevFiles) => {
      const updatedFiles = prevFiles.filter((_, i) => i !== index);
      // Pass the updated files list to the parent component
      onFileChange(updatedFiles);
      return updatedFiles;
    });
  };

  const isUploadDisabled = previewURLs.length >= 2;

  return (
    <div className="flex flex-col items-center mt-5">
      <label
        htmlFor="file"
        className={`flex items-center justify-center w-full h-20 border-2 border-dashed border-gray-300 rounded-lg cursor-pointer transition-all p-11 ${
          isUploadDisabled
            ? 'cursor-not-allowed opacity-50'
            : 'hover:border-blue-500'
        }`}
      >
        <IoCamera size={24} className="text-blue-800/70" />
        <Text className="ml-2" text="Upload Image" />
      </label>

      <Input
        aria-label="Upload Avatar"
        type="file"
        id="file"
        className="hidden"
        accept="image/*"
        multiple
        onChange={handleChangeFiles}
        disabled={isUploadDisabled}
      />

      {errorMessage && (
        <p className="text-red-500 text-sm mt-2">{errorMessage}</p>
      )}

      <div className="grid grid-cols-2 gap-4 mt-4">
        {previewURLs.map((url, index) => (
          <div key={index} className="relative w-24 h-24">
            <ImageFallback
              src={url}
              alt={`Uploaded image ${index + 1}`}
              width={96}
              height={96}
              className="rounded-md object-cover w-full h-full"
            />
            <Button
              className="absolute top-0 right-0 p-1 !border-none rounded-none !bg-transparent dark:!bg-transparent !text-pink-500"
              onClick={() => handleRemoveImage(index)}
            >
              <IoClose size={16} />
            </Button>
          </div>
        ))}
      </div>
    </div>
  );
};

export default AvatarUploadMultiple;
