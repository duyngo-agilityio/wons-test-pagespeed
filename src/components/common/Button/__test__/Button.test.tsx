import { ButtonProps } from '@nextui-org/react';

// Components
import { Button } from '@/components';

describe('Button', () => {
  const mockOnClick = jest.fn();
  const renderComponent = (props?: ButtonProps) =>
    testLibJestUtils.render(
      <Button onClick={mockOnClick} {...props}>
        Test
      </Button>,
    );

  afterEach(() => {
    mockOnClick.mockClear();
  });

  it('match snapshot', () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });

  it('calls click', () => {
    const { getByText } = renderComponent();

    testLibJestUtils.fireEvent.click(getByText('Test'));

    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('not call click', () => {
    const { getByText } = renderComponent({
      isDisabled: true,
    });

    testLibJestUtils.fireEvent.click(getByText('Test'));

    expect(mockOnClick).not.toHaveBeenCalled();
  });
});
