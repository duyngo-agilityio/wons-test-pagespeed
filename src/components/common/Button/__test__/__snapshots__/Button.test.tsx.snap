// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button match snapshot 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <script>
      !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&false)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}else{c.add('light')}if(e==='light'||e==='dark'||!e)d.style.colorScheme=e||'light'}catch(e){}}()
    </script>
    <button
      class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default data-[hover=true]:opacity-hover border-transparent !bg-gray-50 dark:bg-gray-600 text-blue-800 dark:text-white hover:!bg-gray-200/50 dark:hover:!bg-gray-900 px-5 pb-2.5 pt-[11px] text-md rounded-[10px] border-[1px] h-auto min-w-max"
      type="button"
    >
      Test
    </button>
  </div>
</div>
`;
